@claude 
非全屏无限闪烁的问题出现了，一旦全屏又停止闪烁并稳定，再次小窗口，不再闪烁。
后端观察到 反复重复
[INFO] 2026-04-16 14:09:07 [fsBridge] scanForSkills finished. Found 19 skills.
[INFO] 2026-04-16 14:09:07 [WorkerManage] getTaskByIdRollbackBuild: id=88bbb7bc, options=undefined
[INFO] 2026-04-16 14:09:07 [WorkerManage] Found existing task in memory for: 88bbb7bc
[INFO] 2026-04-16 14:09:07 [OpenClawBridge] Model selected {"conversationId":"88bbb7bc","modelId":"gemini-3-flash-preview","modelRatio":3.5,"timestamp":"2026-04-16T06:09:07.575Z"}
[INFO] 2026-04-16 14:09:07 [WorkerManage] getTaskByIdRollbackBuild: id=88bbb7bc, options=undefined
[INFO] 2026-04-16 14:09:07 [WorkerManage] Found existing task in memory for: 88bbb7bc
[ERROR] 2026-04-16 14:09:07 [fsBridge] Failed to get file metadata: {"name":"Error","message":"ENOENT: no such file or directory, stat 'C:\\Users\\16674\\.nexus\\channel-media\\wechat\\wechat_1776145160049_8ouuj1.jpg'","stack":"Error: ENOENT: no such file or directory, stat 'C:\\Users\\16674\\.nexus\\channel-media\\wechat\\wechat_1776145160049_8ouuj1.jpg'"}
[INFO] 2026-04-16 14:09:09 [OpenClawBridge] Model selected {"conversationId":"88bbb7bc","modelId":"gemini-3-flash-preview","modelRatio":3.5,"timestamp":"2026-04-16T06:09:09.718Z"}
[INFO] 2026-04-16 14:09:09 [WorkerManage] getTaskByIdRollbackBuild: id=88bbb7bc, options=undefined
[INFO] 2026-04-16 14:09:09 [WorkerManage] Found existing task in memory for: 88bbb7bc
[INFO] 2026-04-16 14:09:09 [fsBridge] scanForSkills called with path: C:\Users\16674\.nexus\channel-media\wechat/skills
[INFO] 2026-04-16 14:09:09 [FileWatch] Started directory watch id=0e12161b-ec8b-4702-8b6b-72bb2025ebf4 dir=C:\Users\16674\.nexus\channel-media\wechat recursive=true subWatchers=1
[INFO] 2026-04-16 14:09:09 [fsBridge] scanForSkills finished. Found 19 skills.
[INFO] 2026-04-16 14:09:10 [OpenClawBridge] Model selected {"conversationId":"88bbb7bc","modelId":"gemini-3-flash-preview","modelRatio":3.5,"timestamp":"2026-04-16T06:09:10.233Z"}
[INFO] 2026-04-16 14:09:10 [WorkerManage] getTaskByIdRollbackBuild: id=88bbb7bc, options=undefined
[INFO] 2026-04-16 14:09:10 [WorkerManage] Found existing task in memory for: 88bbb7bc
[INFO] 2026-04-16 14:09:10 [WorkerManage] getTaskByIdRollbackBuild: id=88bbb7bc, options=undefined
[INFO] 2026-04-16 14:09:10 [WorkerManage] Found existing task in memory for: 88bbb7bc
[INFO] 2026-04-16 14:09:10 [WorkerManage] getTaskByIdRollbackBuild: id=88bbb7bc, options=undefined
[INFO] 2026-04-16 14:09:10 [WorkerManage] Found existing task in memory for: 88bbb7bc
[INFO] 2026-04-16 14:09:10 [OpenClawBridge] Model selected {"conversationId":"88bbb7bc","modelId":"gemini-3-flash-preview","modelRatio":3.5,"timestamp":"2026-04-16T06:09:10.504Z"}
[INFO] 2026-04-16 14:09:10 [WorkerManage] getTaskByIdRollbackBuild: id=88bbb7bc, options=undefined
[INFO] 2026-04-16 14:09:10 [WorkerManage] Found existing task in memory for: 88bbb7bc
[INFO] 2026-04-16 14:09:10 [fsBridge] scanForSkills called with path: C:\Users\16674\.nexus\channel-media\wechat/skills
[INFO] 2026-04-16 14:09:10 [FileWatch] Started directory watch id=d654de08-58d0-4f48-ac32-fd04bf36726c dir=C:\Users\16674\.nexus\channel-media\wechat recursive=true subWatchers=1
[INFO] 2026-04-16 14:09:10 [fsBridge] scanForSkills finished. Found 19 skills.
[INFO] 2026-04-16 14:09:10 [WorkerManage] getTaskByIdRollbackBuild: id=88bbb7bc, options=undefined
[INFO] 2026-04-16 14:09:10 [WorkerManage] Found existing task in memory for: 88bbb7bc
[INFO] 2026-04-16 14:09:10 [OpenClawBridge] Model selected {"conversationId":"88bbb7bc","modelId":"gemini-3-flash-preview","modelRatio":3.5,"timestamp":"2026-04-16T06:09:10.953Z"}
[INFO] 2026-04-16 14:09:11 [WorkerManage] getTaskByIdRollbackBuild: id=88bbb7bc, options=undefined
[INFO] 2026-04-16 14:09:11 [WorkerManage] Found existing task in memory for: 88bbb7bc
[ERROR] 2026-04-16 14:09:11 [fsBridge] Failed to get file metadata: {"name":"Error","message":"ENOENT: no such file or directory, stat 'C:\\Users\\16674\\.nexus\\channel-media\\wechat\\wechat_1776145160049_8ouuj1.jpg'","stack":"Error: ENOENT: no such file or directory, stat 'C:\\Users\\16674\\.nexus\\channel-media\\wechat\\wechat_1776145160049_8ouuj1.jpg'"}
[INFO] 2026-04-16 14:09:14 [OpenClawBridge] Model selected {"conversationId":"88bbb7bc","modelId":"gemini-3-flash-preview","modelRatio":3.5,"timestamp":"2026-04-16T06:09:14.495Z"}
[INFO] 2026-04-16 14:09:14 [WorkerManage] getTaskByIdRollbackBuild: id=88bbb7bc, options=undefined
[INFO] 2026-04-16 14:09:14 [WorkerManage] Found existing task in memory for: 88bbb7bc
[INFO] 2026-04-16 14:09:14 [fsBridge] scanForSkills called with path: C:\Users\16674\.nexus\channel-media\wechat/skills
[INFO] 2026-04-16 14:09:14 [FileWatch] Started directory watch id=eaedf186-d7a2-479e-bfa8-b8d2b553e596 dir=C:\Users\16674\.nexus\channel-media\wechat recursive=true subWatchers=1
[INFO] 2026-04-16 14:09:14 [fsBridge] scanForSkills finished. Found 19 skills.
[INFO] 2026-04-16 14:09:14 [WorkerManage] getTaskByIdRollbackBuild: id=88bbb7bc, options=undefined
[INFO] 2026-04-16 14:09:14 [WorkerManage] Found existing task in memory for: 88bbb7bc
[INFO] 2026-04-16 14:09:14 [OpenClawBridge] Model selected {"conversationId":"88bbb7bc","modelId":"gemini-3-flash-preview","modelRatio":3.5,"timestamp":"2026-04-16T06:09:14.993Z"}

前端则有一些warn循环
useWorkspaceTree.ts:142 [WS_DEBUG] getWorkspace returned 
WS_DEBUG] loadWorkspace called

--- a/src/renderer/components/FilePreview.tsx
+++ b/src/renderer/components/FilePreview.tsx
@@ -47,19 +47,31 @@ const FilePreview: React.FC<FilePreviewProps> = ({ path, onRemove, readonly = fa
   const fileExt = getFileExtension(path).toUpperCase().replace('.', '');
   const [imageUrl, setImageUrl] = useState<string>('');
   const [fileSize, setFileSize] = useState<string>('');
+  // Track whether the file was not found (ENOENT) so we don't retry on
+  // remount and avoid spamming the backend with repeated IPC calls for a
+  // file that doesn't exist.
+  const [fileError, setFileError] = useState(false);
 
   useEffect(() => {
     // bdpan:// paths are remote — skip local fs operations
     if (path.startsWith('bdpan://')) return;
 
+    // Reset error state when path changes
+    setFileError(false);
+
+    let cancelled = false;
+
     // 获取文件大小
     ipcBridge.fs.getFileMetadata
       .invoke({ path })
       .then((metadata) => {
-        setFileSize(formatFileSize(metadata.size));
+        if (!cancelled) setFileSize(formatFileSize(metadata.size));
       })
       .catch((error) => {
-        console.error('[FilePreview] Failed to get file metadata:', { path, error });
+        if (!cancelled) {
+          console.error('[FilePreview] Failed to get file metadata:', { path, error });
+          setFileError(true);
+        }
       });
 
     // 如果是图片，获取图片的base64
@@ -68,12 +80,19 @@ const FilePreview: React.FC<FilePreviewProps> = ({ path, onRemove, readonly = fa
       ipcBridge.fs.getImageBase64
         .invoke({ path })
         .then((base64) => {
-          setImageUrl(base64);
+          if (!cancelled) setImageUrl(base64);
         })
         .catch((error) => {
-          console.error('[FilePreview] Failed to load image:', { path, error });
+          if (!cancelled) {
+            console.error('[FilePreview] Failed to load image:', { path, error });
+            setFileError(true);
+          }
         });
     }
+
+    return () => {
+      cancelled = true;
+    };
   }, [path, isImage]);
 
   const handleRemove = (e: React.MouseEvent) => {

--- a/src/renderer/layout.tsx
+++ b/src/renderer/layout.tsx
@@ -13,7 +13,7 @@ import Titlebar from '@/renderer/components/Titlebar';
 import { Layout as ArcoLayout } from '@arco-design/web-react';
 import { MenuFold, MenuUnfold } from '@icon-park/react';
 import classNames from 'classnames';
-import React, { Suspense, useCallback, useEffect, useRef, useState } from 'react';
+import React, { Suspense, useCallback, useEffect, useMemo, useRef, useState } from 'react';
 import { Outlet, useLocation } from 'react-router-dom';
 import { LayoutContext } from './context/LayoutContext';
 import { useDeepLink } from './hooks/useDeepLink';
@@ -269,8 +269,17 @@ const Layout: React.FC<{
     collapsedRef.current = collapsed;
   }, [collapsed]);
 
+  // Memoize the context value to prevent unnecessary re-renders of all
+  // LayoutContext consumers when unrelated state (e.g. viewportWidth) changes.
+  // Without this, every resize event creates a new object reference, causing
+  // all consumers (including ChatWorkspace) to re-render.
+  const layoutContextValue = useMemo(
+    () => ({ isMobile, siderCollapsed: collapsed, setSiderCollapsed: setCollapsed }),
+    [isMobile, collapsed, setCollapsed]
+  );
+
   return (
-    <LayoutContext.Provider value={{ isMobile, siderCollapsed: collapsed, setSiderCollapsed: setCollapsed }}>
+    <LayoutContext.Provider value={layoutContextValue}>
       <div className='app-shell relative flex flex-col size-full min-h-0'>
         <Titlebar workspaceAvailable={workspaceAvailable} />
         {/* 移动端左侧边栏蒙板 / Mobile left sider backdrop */}

--- a/src/renderer/pages/conversation/workspace/WorkspaceSkills.tsx
+++ b/src/renderer/pages/conversation/workspace/WorkspaceSkills.tsx
@@ -345,12 +345,19 @@ const WorkspaceSkills = React.forwardRef<WorkspaceSkillsHandle, WorkspaceSkillsP
   loadingCallbackRef.current = onLoadingChange;
   syncedCallbackRef.current = onSynced;
 
+  // Minimum interval between dirChanged-triggered scans to prevent feedback
+  // loops on Windows where file reads during scanForSkills can trigger
+  // further fs.watch events, causing infinite re-renders.
+  const lastScanAtRef = useRef(0);
+  const SCAN_COOLDOWN_MS = 2000;
+
   const scanBoth = useCallback(async () => {
     if (!workspace) {
       setSkills([]);
       return;
     }
     const mySeq = ++reqSeqRef.current;
+    lastScanAtRef.current = Date.now();
     setLoading(true);
     loadingCallbackRef.current?.(true);
     try {
@@ -387,16 +394,27 @@ const WorkspaceSkills = React.forwardRef<WorkspaceSkillsHandle, WorkspaceSkillsP
   // file tree listens to. We don't need a separate watcher — the workspace
   // watcher already covers `skills/` and `.claude/skills/`.
   //
-  // When watchIdRef is provided, only react to events from the workspace
-  // watcher — not all watchers globally. This prevents a feedback loop where
-  // scanForSkills file reads trigger further dirChanged events.
+  // IMPORTANT: Block ALL events when watchIdRef is not yet set — mirrors the
+  // guard in workspace/index.tsx. The previous condition
+  // `if (watchIdRef?.current && ...)` had an inverted check that passed
+  // ALL events through when watchIdRef was null, causing an infinite loop
+  // during initial mount (before the workspace watcher resolved).
+  //
+  // Additionally, enforce a minimum interval between dirChanged-triggered
+  // scans to prevent feedback loops on Windows where file system reads
+  // (stat, readdir, readFile) during scanForSkills can spuriously trigger
+  // further fs.watch events.
   useEffect(() => {
     const unsubscribe = ipcBridge.fileWatch.dirChanged.on((payload) => {
-      // Filter by workspace watchId when available to avoid global feedback loop
-      if (watchIdRef?.current && payload.watchId !== watchIdRef.current) return;
+      // Block events when the workspace watcher hasn't been set up yet, or
+      // when the event comes from a different watcher (e.g. one created by
+      // a scanForSkills call on the backend).
+      if (!watchIdRef?.current || payload.watchId !== watchIdRef.current) return;
       if (debounceRef.current) clearTimeout(debounceRef.current);
       debounceRef.current = setTimeout(() => {
         debounceRef.current = null;
+        // Skip if a scan ran recently — prevents the scan → dirChanged → scan loop
+        if (Date.now() - lastScanAtRef.current < SCAN_COOLDOWN_MS) return;
         void scanBoth();
       }, 250);
     });

--- a/src/renderer/pages/conversation/workspace/WorkspaceSkills.tsx
+++ b/src/renderer/pages/conversation/workspace/WorkspaceSkills.tsx
@@ -58,6 +58,12 @@ export interface WorkspaceSkillsProps {
   onLoadingChange?: (loading: boolean) => void;
   /** Notifies the parent that a refresh cycle just finished. */
   onSynced?: () => void;
+  /**
+   * Ref to the workspace directory watcher ID. When provided, dirChanged
+   * events are filtered to only this watcher — preventing a global feedback
+   * loop where scanForSkills file reads trigger more dirChanged events.
+   */
+  watchIdRef?: React.RefObject<string | null>;
 }
 
 export interface WorkspaceSkillsHandle {
@@ -328,7 +334,7 @@ const SkillIconGraphic: React.FC<{
   return <Icon theme='outline' size='16' fill={fillColor} />;
 };
 
-const WorkspaceSkills = React.forwardRef<WorkspaceSkillsHandle, WorkspaceSkillsProps>(({ workspace, eventPrefix, backend, searchQuery, onLoadingChange, onSynced }, ref) => {
+const WorkspaceSkills = React.forwardRef<WorkspaceSkillsHandle, WorkspaceSkillsProps>(({ workspace, eventPrefix, backend, searchQuery, onLoadingChange, onSynced, watchIdRef }, ref) => {
   const { t } = useTranslation();
   const [skills, setSkills] = useState<SkillItem[]>([]);
   const [loading, setLoading] = useState(false);
@@ -380,8 +386,14 @@ const WorkspaceSkills = React.forwardRef<WorkspaceSkillsHandle, WorkspaceSkillsP
   // inotify-style auto-refresh: piggyback on the same `dirChanged` stream the
   // file tree listens to. We don't need a separate watcher — the workspace
   // watcher already covers `skills/` and `.claude/skills/`.
+  //
+  // When watchIdRef is provided, only react to events from the workspace
+  // watcher — not all watchers globally. This prevents a feedback loop where
+  // scanForSkills file reads trigger further dirChanged events.
   useEffect(() => {
-    const unsubscribe = ipcBridge.fileWatch.dirChanged.on(() => {
+    const unsubscribe = ipcBridge.fileWatch.dirChanged.on((payload) => {
+      // Filter by workspace watchId when available to avoid global feedback loop
+      if (watchIdRef?.current && payload.watchId !== watchIdRef.current) return;
       if (debounceRef.current) clearTimeout(debounceRef.current);
       debounceRef.current = setTimeout(() => {
         debounceRef.current = null;

--- a/src/renderer/pages/conversation/workspace/hooks/useWorkspaceEvents.ts
+++ b/src/renderer/pages/conversation/workspace/hooks/useWorkspaceEvents.ts
@@ -39,6 +39,10 @@ interface UseWorkspaceEventsOptions {
 /**
  * useWorkspaceEvents - 管理所有事件监听器
  * Manage all event listeners
+ *
+ * Returns { watchIdRef } so sibling hooks (skillCount, WorkspaceSkills) can
+ * filter dirChanged events by the same watchId instead of reacting to every
+ * file-system event globally.
  */
 export function useWorkspaceEvents(options: UseWorkspaceEventsOptions) {
   const { conversation_id, eventPrefix, workspace, refreshWorkspace, clearSelection, setFiles, setSelected, setExpandedKeys, setTreeKey, selectedNodeRef, selectedKeysRef, closeContextMenu, setContextMenu, closeRenameModal, closeDeleteModal } = options;
@@ -51,6 +55,10 @@ export function useWorkspaceEvents(options: UseWorkspaceEventsOptions) {
   useEffect(() => {
     refreshRef.current = refreshWorkspace;
   }, [refreshWorkspace]);
+
+  // Expose the active directory-watch ID so other hooks can filter dirChanged
+  // events to only the workspace watcher, avoiding a global feedback loop.
+  const watchIdRef = useRef<string | null>(null);
 
   /**
    * 监听对话切换事件 - 重置所有状态
@@ -106,7 +114,7 @@ export function useWorkspaceEvents(options: UseWorkspaceEventsOptions) {
     if (!workspace) return;
 
     let cancelled = false;
-    let activeWatchId: string | null = null;
+    let localWatchId: string | null = null;
     let debounceTimer: ReturnType<typeof setTimeout> | null = null;
 
     const scheduleRefresh = () => {
@@ -122,7 +130,7 @@ export function useWorkspaceEvents(options: UseWorkspaceEventsOptions) {
     };
 
     const unsubscribe = ipcBridge.fileWatch.dirChanged.on((payload) => {
-      if (!activeWatchId || payload.watchId !== activeWatchId) return;
+      if (!localWatchId || payload.watchId !== localWatchId) return;
       scheduleRefresh();
     });
 
@@ -136,7 +144,8 @@ export function useWorkspaceEvents(options: UseWorkspaceEventsOptions) {
           return;
         }
         if (res?.success && res.data?.watchId) {
-          activeWatchId = res.data.watchId;
+          localWatchId = res.data.watchId;
+          watchIdRef.current = res.data.watchId;
         }
       } catch {
         /* workspace may not exist yet; ignore, agent events will still refresh */
@@ -150,10 +159,11 @@ export function useWorkspaceEvents(options: UseWorkspaceEventsOptions) {
         debounceTimer = null;
       }
       unsubscribe();
-      if (activeWatchId) {
-        ipcBridge.fileWatch.stopWatchDir.invoke({ watchId: activeWatchId }).catch(() => {});
-        activeWatchId = null;
+      if (localWatchId) {
+        ipcBridge.fileWatch.stopWatchDir.invoke({ watchId: localWatchId }).catch(() => {});
+        localWatchId = null;
       }
+      watchIdRef.current = null;
     };
   }, [workspace]);
 
@@ -244,4 +254,6 @@ export function useWorkspaceEvents(options: UseWorkspaceEventsOptions) {
       window.removeEventListener('keydown', handleKeyDown);
     };
   }, [closeContextMenu]);
+
+  return { watchIdRef };
 }

--- a/src/renderer/pages/conversation/workspace/hooks/useWorkspaceEvents.ts
+++ b/src/renderer/pages/conversation/workspace/hooks/useWorkspaceEvents.ts
@@ -110,12 +110,18 @@ export function useWorkspaceEvents(options: UseWorkspaceEventsOptions) {
    * differences (macOS/Windows native recursive vs. Linux per-dir walk) are
    * handled transparently by the main-process bridge.
    */
+  // Minimum interval between dirChanged-triggered refreshes to prevent
+  // feedback loops on Windows where file system reads during getWorkspace
+  // (stat, readdir) can spuriously trigger further fs.watch events.
+  const lastDirRefreshAtRef = useRef(0);
+
   useEffect(() => {
     if (!workspace) return;
 
     let cancelled = false;
     let localWatchId: string | null = null;
     let debounceTimer: ReturnType<typeof setTimeout> | null = null;
+    const DIR_REFRESH_COOLDOWN_MS = 1000;
 
     const scheduleRefresh = () => {
       if (debounceTimer) clearTimeout(debounceTimer);
@@ -125,6 +131,10 @@ export function useWorkspaceEvents(options: UseWorkspaceEventsOptions) {
       debounceTimer = setTimeout(() => {
         debounceTimer = null;
         if (cancelled) return;
+        // Skip if we refreshed recently — prevents the refresh → file reads
+        // → dirChanged → refresh loop that occurs on Windows.
+        if (Date.now() - lastDirRefreshAtRef.current < DIR_REFRESH_COOLDOWN_MS) return;
+        lastDirRefreshAtRef.current = Date.now();
         refreshRef.current();
       }, 200);
     };

--- a/src/renderer/pages/conversation/workspace/index.tsx
+++ b/src/renderer/pages/conversation/workspace/index.tsx
@@ -208,6 +208,15 @@ const ChatWorkspace: React.FC<WorkspaceProps> = ({ conversation_id, workspace, e
 
   // Skill count for the skills tab pill — scanned from the workspace skill
   // root that belongs to the current conversation backend.
+  //
+  // IMPORTANT: We filter dirChanged events by the workspace watchId to avoid a
+  // feedback loop where scanForSkills file-system reads trigger further
+  // dirChanged events, which trigger more scanForSkills calls, causing
+  // infinite flickering in non-fullscreen windows.
+  //
+  // NOTE: `watchIdRef` is defined later (from useWorkspaceEvents) but this is
+  // safe because useEffect callbacks run after the entire component body, so
+  // watchIdRef is always initialized before the listener fires.
   const [skillCount, setSkillCount] = useState(0);
   useEffect(() => {
     if (!workspace) {
@@ -227,7 +236,11 @@ const ChatWorkspace: React.FC<WorkspaceProps> = ({ conversation_id, workspace, e
     };
     void refreshCount();
     let debounce: ReturnType<typeof setTimeout> | null = null;
-    const unsubscribe = ipcBridge.fileWatch.dirChanged.on(() => {
+    const unsubscribe = ipcBridge.fileWatch.dirChanged.on((payload) => {
+      // Only react to events from our workspace watcher — not all watchers
+      // globally. This prevents the scanForSkills → file read → dirChanged
+      // feedback loop that caused infinite re-renders.
+      if (!watchIdRef.current || payload.watchId !== watchIdRef.current) return;
       if (debounce) clearTimeout(debounce);
       debounce = setTimeout(() => {
         debounce = null;
@@ -428,8 +441,10 @@ const ChatWorkspace: React.FC<WorkspaceProps> = ({ conversation_id, workspace, e
     openPreview,
   });
 
-  // Setup events
-  useWorkspaceEvents({
+  // Setup events — capture the watchIdRef so the skillCount effect below can
+  // filter dirChanged events to only the workspace watcher, preventing a
+  // global feedback loop (scanForSkills → file reads → dirChanged → repeat).
+  const { watchIdRef } = useWorkspaceEvents({
     conversation_id,
     eventPrefix,
     workspace,
@@ -1130,7 +1145,7 @@ const ChatWorkspace: React.FC<WorkspaceProps> = ({ conversation_id, workspace, e
         {/* Main content area — Skills grid OR Search + Tree. */}
         {activeTab === 'skills' && (
           <FlexFullContainer containerClassName='workspace-card__body workspace-card__body--skills overflow-y-auto'>
-            <WorkspaceSkills ref={skillsHandleRef} workspace={workspace} eventPrefix={eventPrefix} backend={backend} searchQuery={searchText} onLoadingChange={setSkillsLoading} onSynced={() => setLastSyncAt(Date.now())} />
+            <WorkspaceSkills ref={skillsHandleRef} workspace={workspace} eventPrefix={eventPrefix} backend={backend} searchQuery={searchText} onLoadingChange={setSkillsLoading} onSynced={() => setLastSyncAt(Date.now())} watchIdRef={watchIdRef} />
           </FlexFullContainer>
         )}
 

--- a/src/renderer/pages/conversation/workspace/index.tsx
+++ b/src/renderer/pages/conversation/workspace/index.tsx
@@ -214,17 +214,25 @@ const ChatWorkspace: React.FC<WorkspaceProps> = ({ conversation_id, workspace, e
   // dirChanged events, which trigger more scanForSkills calls, causing
   // infinite flickering in non-fullscreen windows.
   //
+  // Additionally, we enforce a minimum interval between dirChanged-triggered
+  // scans to prevent feedback loops on Windows where file-system reads (stat,
+  // readdir) during scanForSkills can spuriously trigger further fs.watch
+  // events on the same directory tree.
+  //
   // NOTE: `watchIdRef` is defined later (from useWorkspaceEvents) but this is
   // safe because useEffect callbacks run after the entire component body, so
   // watchIdRef is always initialized before the listener fires.
   const [skillCount, setSkillCount] = useState(0);
+  const lastSkillScanAtRef = useRef(0);
   useEffect(() => {
     if (!workspace) {
       setSkillCount(0);
       return undefined;
     }
     let cancelled = false;
+    const SKILL_SCAN_COOLDOWN_MS = 2000;
     const refreshCount = async () => {
+      lastSkillScanAtRef.current = Date.now();
       try {
         const skillRoot = resolveWorkspaceSkillRoot(workspace, eventPrefix, backend);
         const result = await ipcBridge.fs.scanForSkills.invoke({ folderPath: skillRoot.path }).catch((): undefined => undefined);
@@ -244,6 +252,8 @@ const ChatWorkspace: React.FC<WorkspaceProps> = ({ conversation_id, workspace, e
       if (debounce) clearTimeout(debounce);
       debounce = setTimeout(() => {
         debounce = null;
+        // Skip if a scan ran recently — prevents the scan → dirChanged → scan loop
+        if (Date.now() - lastSkillScanAtRef.current < SKILL_SCAN_COOLDOWN_MS) return;
         void refreshCount();
       }, 250);
     });

